### PR TITLE
Scaleway Database Backup : Create new module

### PR DESCRIPTION
**Pull request by guillaume-ro-fr**
##### SUMMARY
New module for Scaleway provider : [Managed database backups](https://developers.scaleway.com/en/products/rdb/api/#backups-d14bc7)

##### ISSUE TYPE
New Module Pull Request

##### COMPONENT NAME
`scaleway_database_backup` : Create, edit, delete, export and restore a backup

### CODE REVIEW

**Comment by guillaume-ro-fr** [comment]
Error from Shippable: 

`ERROR: plugins/modules/cloud/scaleway/scaleway_database_backup.py:0:0: doc-required-mismatch: Argument 'api_token' in argument_spec is required, but is not documented as being required (75%)`

Even with the code below, `api_token` is not detected in documentation: 

```yaml
extends_documentation_fragment:
- community.general.scaleway
```

File: plugins/modules/scaleway_database_backup.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
+#!/usr/bin/python
```
Comment (felixfontein) [inline plugins/modules/scaleway_database_backup.py:1]: This does not need to be a copy of the module, but a symlink to the real module.

**Comment by felixfontein** [comment]
Looks like there is a bug in the docs fragment which has been simply ignored for the other scaleway modules.

File: plugins/modules/scaleway_database_backup.py (hunk 1 above)
Comment (guillaume-ro-fr) [reply plugins/modules/scaleway_database_backup.py]: Hmmmm, I think symlinks don't work, [the tests](https://dashboard.zuul.ansible.com/t/ansible/build/f2e8156e56ad4d46b490695d13d5c8dd/console) fail since I deleted the file and replaced it with a relative symlink.

Comment (guillaume-ro-fr) [reply plugins/modules/scaleway_database_backup.py]: Resolved - My bad, the link was broken

**Comment by guillaume-ro-fr** [comment]
Error in CI: `changelogs/fragments/.empty:0:0: file must not be a dotfile`

**Comment by felixfontein** [comment]
Please don't fix the problem with the other scaleway modules in this PR, create a separate PR for that.

About the remaining error: I'll fix that.

**Comment by guillaume-ro-fr** [comment]
Scaleway `api_token` correction moved to #747

**Comment by felixfontein** [comment]
I restarted the tests, so they run against the updated `main` branch.

**Comment by Andersson007** [comment]
@felixfontein do we need a changelog fragment here?

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: scaleway_database_backup
+short_description: Scaleway database backups management module
+author: Guillaume Rodriguez (@guillaume_ro_fr)
+description:
+    - This module manages database backups on Scaleway account
+      U(https://developer.scaleway.com)
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:20]:
```suggestion
    - This module manages database backups on Scaleway account U(https://developer.scaleway.com).
```
I'd move the reference to `seealso:` section. Anyway the dot should be there (we shouldn't put it in `short_description` but should in `description`)

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+    default: present
+    choices:
+        - present
+        - absent
+        - exported
+        - restored
+
+  region:
+    description:
+        - Scaleway region to use (for example fr-par).
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:37]:
```suggestion
        - Scaleway region to use (for example C(fr-par)).
```

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+    choices:
+        - fr-par
+        - nl-ams
+
+  id:
+    description:
+        - UUID used to identify the database backup.
+        - Required for C(absent), C(exported) and C(restored) states
+    type: str
+    required: false
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:49]:
```suggestion
```
this is the default value. Could you please remove `required: false` from everywhere?

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+    description:
+        - Name used to identify the database backup.
+        - Required for C(present) state
+    type: str
+    required: false
+
+  database_name:
+    description:
+        - Name used to identify the database.
+        - Required for C(present) and C(restored) states
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:61]:
```suggestion
        - Required for C(present) and C(restored) states.
```

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+    description:
+        - UUID used to identify the database backup.
+        - Required for C(absent), C(exported) and C(restored) states
+    type: str
+    required: false
+
+  name:
+    description:
+        - Name used to identify the database backup.
+        - Required for C(present) state
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:54]:
```suggestion
        - Required for C(present) state.
```

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+    type: str
+    required: true
+    choices:
+        - fr-par
+        - nl-ams
+
+  id:
+    description:
+        - UUID used to identify the database backup.
+        - Required for C(absent), C(exported) and C(restored) states
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:47]:
```suggestion
        - Required for C(absent), C(exported) and C(restored) states.
```

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+    description:
+        - Name used to identify the database.
+        - Required for C(present) and C(restored) states
+    type: str
+    required: false
+
+  instance_id:
+    description:
+        - UUID of the instance associated to the database backup
+        - Required for C(present) and C(restored) states
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:68]:
```suggestion
        - UUID of the instance associated to the database backup.
        - Required for C(present) and C(restored) states.
```

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+    type: str
+    required: false
+
+  expires_at:
+    description:
+        - Expiration datetime of the database backup (ISO 8601 format)
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:74]:
```suggestion
        - Expiration datetime of the database backup (ISO 8601 format).
```

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+
+  wait:
+    description:
+        - Wait for the instance to reach its desired state before returning.
+    type: bool
+    default: 'no'
+
+  wait_timeout:
+    description:
+        - Time to wait for the backup to reach the expected state
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:86]:
```suggestion
        - Time to wait for the backup to reach the expected state.
```

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+    type: int
+    required: false
+    default: 300
+
+  wait_sleep_time:
+    description:
+        - Time to wait before every attempt to check the state of the backup
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:93]:
```suggestion
        - Time to wait before every attempt to check the state of the backup.
```

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+    type: int
+    required: false
+    default: 3
+'''
+
+EXAMPLES = '''
+  - name: Create a backup
+    scaleway_database_backup:
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:101]:
```suggestion
    community.general.scaleway_database_backup:
```

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+        name: 'my_backup'
+        state: present
+        region: 'fr-par'
+        database_name: 'my-database'
+        instance_id: '50968a80-2909-4e5c-b1af-a2e19860dddb'
+
+  - name: Export a backup
+    scaleway_database_backup:
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:109]:
```suggestion
    community.general.scaleway_database_backup:
```

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+        id: '6ef1125a-037e-494f-a911-6d9c49a51691'
+        state: exported
+        region: 'fr-par'
+
+  - name: Restore backup
+    scaleway_database_backup:
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:115]:
```suggestion
    community.general.scaleway_database_backup:
```

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+        id: '6ef1125a-037e-494f-a911-6d9c49a51691'
+        state: restored
+        region: 'fr-par'
+        database_name: 'my-new-database'
+        instance_id: '50968a80-2909-4e5c-b1af-a2e19860dddb'
+'''
+
+RETURN = '''
+data:
+    description: This is only present when C(state=present)
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:125]:
```suggestion
data:
    description: Backup metadata.
```
imo, it sounds like all data from a backup will be returned:) maybe `backup_info` or `backup_metadata` or something. And we have the info from `description` in `returned` subsection.

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+    }
+'''
+
+import datetime
+import time
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible_collections.community.general.plugins.module_utils.scaleway import (
+    Scaleway, scaleway_argument_spec,
+    SCALEWAY_REGIONS)
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:153]:
```suggestion
    Scaleway,
    scaleway_argument_spec,
    SCALEWAY_REGIONS,
)
```
just a good practice

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+    state_strategy[state](module, account_api, backup_by_id)
+
+
+def main():
+    argument_spec = scaleway_argument_spec()
+    argument_spec.update(dict(
+        state=dict(default='present', choices=['absent', 'present', 'exported', 'restored']),
+        region=dict(required=True, choices=SCALEWAY_REGIONS),
+        id=dict(),
+        name=dict(required=False),
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:329]:
```suggestion
        name=dict(type='str'),
```
it's fair for all the parameters

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 18]
```diff
@@ -0,0 +1,346 @@
...
+        instance_id=dict(required=False),
+        expires_at=dict(),
+        wait=dict(type='bool', default=False),
+        wait_timeout=dict(type='int', default=300),
+        wait_sleep_time=dict(type='int', default=3),
+    ))
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+    )
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:340]:
```suggestion
        supports_check_mode=True,
    )
```
you could restrict possible parameter combinations by using something from https://docs.ansible.com/ansible/latest/reference_appendices/module_utils.html (e.g. `required_if`, `required_together`, `mutually_exclusive`, etc.)
if certain parameters are ignored in some cases, it also makes sense to point it out in the doc, e.g.
```
- Ignored when I(state=absent).
```

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 19]
```diff
@@ -0,0 +1,346 @@
...
+        region: 'fr-par'
+
+  - name: Restore backup
+    scaleway_database_backup:
+        id: '6ef1125a-037e-494f-a911-6d9c49a51691'
+        state: restored
+        region: 'fr-par'
+        database_name: 'my-new-database'
+        instance_id: '50968a80-2909-4e5c-b1af-a2e19860dddb'
+'''
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:121]:
```suggestion
        instance_id: '50968a80-2909-4e5c-b1af-a2e19860dddb'
    
    - name: Remove a backup
      ...
'''
```

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 20]
```diff
@@ -0,0 +1,346 @@
...
+        database_name: 'my-database'
+        instance_id: '50968a80-2909-4e5c-b1af-a2e19860dddb'
+
+  - name: Export a backup
+    scaleway_database_backup:
+        id: '6ef1125a-037e-494f-a911-6d9c49a51691'
+        state: exported
+        region: 'fr-par'
+
+  - name: Restore backup
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:114]:
```suggestion
  - name: Restore a backup
```
as in the other examples

**Comment by Andersson007** [comment]
@felixfontein the tests are `unsupported`. Have we decided anything if we want to see them or don't?

**Comment by felixfontein** [comment]
@Andersson007 no, I put it on the agenda of the community meeting, but we haven't met again so far :)

File: changelogs/fragments/741-scaleway-database-backup-creation.yaml [hunk 21]
```diff
@@ -0,0 +1,3 @@
+---
+minor_changes:
+  - scaleway_database_backup - new module for Scaleway provider Managed database backups (https://developers.scaleway.com/en/products/rdb/api/#backups-d14bc7).
```
Comment (felixfontein) [inline changelogs/fragments/741-scaleway-database-backup-creation.yaml:3]: You can remove the changelog fragment, new modules are automatically announced in the changelog.

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py (hunk 18 above)
Comment (guillaume-ro-fr) [reply plugins/modules/cloud/scaleway/scaleway_database_backup.py]: I used `required_if` and `required_together` (thanks for the tip !)

**Comment by guillaume-ro-fr** [comment]
@Andersson007 : Thank you for your great review !

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 22]
```diff
@@ -0,0 +1,365 @@
...
+        region: 'fr-par'
+'''
+
+RETURN = '''
+data:
+    description: Backup metadata.
+    returned: when C(state=present), C(state=exported) or C(state=restored)
+    type: dict
+    sample: {
+        "data": {
```
Comment (Andersson007) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:137]:
```suggestion
metadata:
    description: Backup metadata.
    returned: when C(state=present), C(state=exported) or C(state=restored)
    type: dict
    sample: {
        "metadata": {
```
maybe the key `data` => `metadata` or `status` or something else?

**Comment by Andersson007** [comment]
> @Andersson007 : Thank you for your great review !

Yw!

**Comment by Andersson007** [comment]
@sieben @hekonsek @abarbare @Spredzy could you please take a look. @felixfontein and I don't work with `scaleway`, so, difficult to suggest something

**Comment by Andersson007** [comment]
@felixfontein do you have anything to add?
if no, and if there's no feedback from the folks until tomorrow, i think we could merge it. What do you think?

File: plugins/module_utils/scaleway.py [hunk 23]
```diff
@@ -120,7 +120,6 @@ def _url_builder(self, path, params):
 
     def send(self, method, path, data=None, headers=None, params=None):
         url = self._url_builder(path=path, params=params)
-        self.warn(url)
```
Comment (felixfontein) [inline plugins/module_utils/scaleway.py:123]: This change should probably have a changelog fragment. It might be better to actually move it to a separate PR. It looks like a bug to me.

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 24]
```diff
@@ -0,0 +1,365 @@
...
+        - Expiration datetime of the database backup (ISO 8601 format).
+        - Ignored when C(state=absent), C(state=exported) or C(state=restored).
+    type: str
+    required: false
+
+  wait:
+    description:
+        - Wait for the instance to reach its desired state before returning.
+    type: bool
+    default: 'no'
```
Comment (felixfontein) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:84]:
```suggestion
    default: false
```

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 25]
```diff
@@ -0,0 +1,365 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: scaleway_database_backup
+short_description: Scaleway database backups management module
```
Comment (felixfontein) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:16]:
```suggestion
short_description: Scaleway database backups management module
version_added: 1.1.0
```

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 26]
```diff
@@ -0,0 +1,365 @@
...
+author: Guillaume Rodriguez (@guillaume_ro_fr)
+description:
+    - This module manages database backups on Scaleway account U(https://developer.scaleway.com).
+extends_documentation_fragment:
+    - community.general.scaleway
+options:
+  state:
+    description:
+        - Indicate desired state of the database backup.
```
Comment (felixfontein) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:25]: It would be good to explain what the different states mean here. To me, it is not clear what all these states mean.

File: plugins/modules/cloud/scaleway/scaleway_database_backup.py [hunk 27]
```diff
@@ -0,0 +1,365 @@
...
+    default: present
+    choices:
+        - present
+        - absent
+        - exported
+        - restored
+
+  region:
+    description:
+        - Scaleway region to use (for example C(fr-par)).
```
Comment (felixfontein) [inline plugins/modules/cloud/scaleway/scaleway_database_backup.py:36]: I think this should be part of the module_utils and the docs_fragment, and not be repeated in every module. But that's not something that should be changed in this PR, so let's keep it for now.

**Comment by Andersson007** [comment]
any updates on this?

**Comment by Andersson007** [comment]
we need to see Shippable green

**Comment by felixfontein** [comment]
Unfortunately there's still no feedback from anyone who actually knows about scaleway, but what the heck... @Andersson007 please merge this if you have no further wishes. I'm happy with the current state.

**Comment by Andersson007** [comment]
@guillaume-ro-fr thanks for the new module!
@felixfontein thanks for reviewing!
merged into main
